OfficeAccessCard editing - backend changes

DavidN016 opened the pull request:
-Created a post /edit, and a helper function for edit alias.
-Created edit mode in pencil/edit button, that edits the alias when clicked.
-Created test files for post /edit with dummy document.

Commit:
Add edit alias functionality for OfficeAccessCard (#1900)

- Implemented a new endpoint to edit card aliases, including validation for input.
- Updated the OfficeAccessCard utility to support alias editing.
- Enhanced the CardReader component to allow users to edit card aliases directly in the UI.
- Added corresponding tests to ensure proper functionality and error handling for the new feature.
- Introduced a new audit log action for alias edits.

Commit:
Add edit alias functionality for OfficeAccessCard (#1900)

- Implemented a new endpoint to edit card aliases, including validation for input.
- Updated the OfficeAccessCard utility to support alias editing.
- Enhanced the CardReader component to allow users to edit card aliases directly in the UI.
- Added corresponding tests to ensure proper functionality and error handling for the new feature.
- Introduced a new audit log action for alias edits.

Commit:
fixed lint issues

Commit:
fixed lint issues

adarshm11 commented on src/Pages/CardReader/CardReader.js:
```diff
@@ -89,12 +91,68 @@ export default function CardReader() {
+    try {
+      const response = await editCardAlias(token, cardId, editedAlias.trim());
+      if (!response.error) {
+        setCards(prevCards =>
+          prevCards.map(card =>
+            card._id === cardId
+              ? { ...card, alias: editedAlias.trim() }
+              : card
```
i think what we should do here is when the alias is successfully edited, we query mongodb for all cards again, so that way this alias isn't only client side

adarshm11 commented on src/Pages/Overview/SVG.js:
```diff
@@ -51,6 +51,15 @@ export function trashcanSymbol(color = 'black') {
   );
 }
 
+export function pencilSymbol(color = '#6b7280') {
+  return (
+    <svg width='20' height='20' viewBox='0 0 24 24' fill='none' stroke={color} strokeWidth='2' strokeLinecap='round' strokeLinejoin='round'>
+      <path d='M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7'/>
+      <path d='m18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z'/>
+    </svg>
+  );
+}
+
```
this doesn't have to be put in this file, the `trashCanSymbol` function is only imported from here because it was already implemented here when card reader page was made. you can move this to CardReader.js

adarshm11 commented on test/api/OfficeAccessCard.js:
```diff
@@ -241,4 +250,101 @@ describe('OfficeAccessCard', () => {
+      expect(result).to.have.status(BAD_REQUEST);
+    });
+
+    it('Should return 400 when both _id and alias are missing', async () => {
+      setTokenStatus(true);
+      const result = await test.sendPostRequestWithToken(token,
+        EDIT_API_PATH, {});
+      expect(result).to.have.status(BAD_REQUEST);
+    });
```
this test is redundant

adarshm11 commented on test/api/OfficeAccessCard.js:
```diff
@@ -241,4 +250,101 @@ describe('OfficeAccessCard', () => {
+      expect(updatedCard.alias).to.equal(NEW_ALIAS);
+    });
+  });
+
```
very impressive how thorough these tests are :)

Commit:
fixed issues with edit alias

Commit:
fixed issues with edit alias

adarshm11 committed:
fixed input re-sizing

adarshm11 committed:
fixed input re-sizing

evanugarte commented on api/main_endpoints/routes/OfficeAccessCard.js:
```diff
@@ -215,6 +216,57 @@ router.post('/getAllCards', async (req, res) => {
 });
 
+router.post('/edit', async (req, res) => {
+  const decoded = decodeToken(req);
+  if (!decoded) {
+    return res.sendStatus(UNAUTHORIZED);
+  }
```
i think we realized this doesnt work anymore, need to merge #1877 first

ill work on it

evanugarte reviewed (commented):
can there be some screenshots of the new ui

adarshm11 commented:
> can there be some screenshots of the new ui

<img width="1041" height="420" alt="image" src="https://github.com/user-attachments/assets/14da0947-3142-4b2a-bfee-17cf4990bff4" />

<img width="1045" height="423" alt="image" src="https://github.com/user-attachments/assets/c6904173-3b6c-4960-a5b7-a07d9bf09f1b" />

adarshm11 committed:
Merge branch 'dev' of https://github.com/SCE-Development/Clark into editalias

adarshm11 committed:
fix token decoding for office cards

adarshm11 committed:
fix token decoding for office cards

adarshm11 committed:
fix failing test

adarshm11 committed:
fix failing test

adarshm11 committed:
separate frontend changes

adarshm11 committed:
separate frontend changes

evanugarte commented on api/main_endpoints/routes/OfficeAccessCard.js:
```diff
@@ -209,6 +218,56 @@ router.post('/getAllCards', async (req, res) => {
+
+  if (!_id || !alias) {
+    return res.status(BAD_REQUEST).send('_id and alias are required in request body');
+  }
```
we should do this approach, from the verify endpoint, alias.trim() can also be in the array

```js
  const required = [
    { value: apiKey, title: 'X-API-Key HTTP header', },
    { value: cardBytes, title: 'cardBytes query parameter', },
  ];
  const missingValue = required.find(({ value }) => !value);
  if (missingValue) {
    writeLogToClient(req.method, {
      statusCode: BAD_REQUEST,
      message: `${missingValue.title} missing from request`
    });
    return res.status(BAD_REQUEST).send(`${missingValue.title} missing from request`);
  }

```

evanugarte commented on api/main_endpoints/routes/OfficeAccessCard.js:
```diff
@@ -209,6 +218,56 @@ router.post('/getAllCards', async (req, res) => {
+  // Validate _id is a valid ObjectId format
+  if (!/^[0-9a-fA-F]{24}$/.test(_id)) {
+    return res.status(BAD_REQUEST).send('_id must be a valid ObjectId');
+  }
```
you could even throw this in the missingFields thing

evanugarte commented on api/main_endpoints/routes/OfficeAccessCard.js:
```diff
@@ -209,6 +218,56 @@ router.post('/getAllCards', async (req, res) => {
+    const updatedCard = await editAlias(_id, alias);
+
+    if (!updatedCard) {
+      return res.status(NOT_FOUND).send('Card not found');
```
we can just do sendStatus, 404 status implies that we didnt find it

adarshm11 replied:
done

evanugarte commented on api/main_endpoints/routes/OfficeAccessCard.js:
```diff
@@ -209,6 +218,56 @@ router.post('/getAllCards', async (req, res) => {
+      action: AuditLogActions.EDIT_CARD,
+      details: {
+        newAlias: alias,
+        oldAlias: updatedCard.alias !== alias ? 'unknown' : alias
```
why would updatedCard.alias not exist

also we could just do 

```
details: {
  newAlias: alias,
  _id,
}
```
no need to say what the old one was

adarshm11 replied:
done

adarshm11 committed:
Merge branch 'editalias' of https://github.com/SCE-Development/Clark into editalias

adarshm11 committed:
modify required checks for /edit

evanugarte commented on api/main_endpoints/routes/OfficeAccessCard.js:
```diff
@@ -80,14 +81,22 @@ router.get('/verify', async (req, res) => {
+      message: 'API key missing from request',
     });
     return res.sendStatus(UNAUTHORIZED);
   }
```
do we even need this? the `required` array already checks this i think

evanugarte commented on api/main_endpoints/routes/OfficeAccessCard.js:
```diff
@@ -209,6 +218,56 @@ router.post('/getAllCards', async (req, res) => {
+  const { _id, alias } = req.body;
+
+  const required = [
+    { value: _id && /^[0-9a-fA-F]{24}$/.test(_id) ? _id : null, title: 'Card ID', },
```
maybe have the title be `valid, alphanumeric Card ID`

evanugarte commented on api/main_endpoints/util/OfficeAccessCard.js:
```diff
@@ -20,8 +20,7 @@ function checkIfCardExists({ cardBytes = null, alias = null } = {}) {
           if (!result) {
-            const { description } = body;
-            logger.info(`Card:${description} not found in the database`);
+            logger.info('Card not found in the database');
```
why did we change this line can we put it back

evanugarte commented on test/api/OfficeAccessCard.js:
```diff
@@ -241,4 +250,94 @@ describe('OfficeAccessCard', () => {
+      expect(result).to.have.status(BAD_REQUEST);
+    });
+
+    it('Should preserve other card properties when updating alias', async () => {
```
up to you, should we add a quick test ensuring that only officers and above can edit a card, not member role

evanugarte commented on api/main_endpoints/routes/OfficeAccessCard.js:
```diff
@@ -209,6 +218,56 @@ router.post('/getAllCards', async (req, res) => {
 });
 
+router.post('/edit', async (req, res) => {
+  const decoded = await decodeToken(req);
```
i think we have to pass in access role officer here, otherwise anyone with a valid token at any role level can do this

adarshm11 committed:
addressed requested changes

adarshm11 committed:
try to fix failing test

adarshm11 committed:
try to fix that failing test one more time

adarshm11 committed:
remove this test cuz it wont work

adarshm11 committed:
OfficeAccessCard editing - UI changes (#1935)

* Add edit alias functionality for OfficeAccessCard (#1900)

- Implemented a new endpoint to edit card aliases, including validation for input.
- Updated the OfficeAccessCard utility to support alias editing.
- Enhanced the CardReader component to allow users to edit card aliases directly in the UI.
- Added corresponding tests to ensure proper functionality and error handling for the new feature.
- Introduced a new audit log action for alias edits.

* fixed lint issues

* fixed issues with edit alias

* fixed input re-sizing

* fix token decoding for office cards

* fix failing test

* card reader frontend

* rebase

* undo diff

* fixed requested changes

* fixed requested changes

---------

Co-authored-by: DavidN016 <david.a.nguyen01@sjsu.edu>
